controllers/routemap.go

```diff
@@ -0,0 +1,420 @@
+package controllers
+
+import (
+	"bytes"
+	"errors"
+	"html/template"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/iter8-tools/iter8/base/log"
+	"github.com/iter8-tools/iter8/controllers/k8sclient"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	seriyaml "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+/* types: begin */
+
+type routemap struct {
+	// Todo: prune this down to agra.ObjectMeta instead of metav1.ObjectMeta
+	mutex             sync.RWMutex
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Variants          []variant                  `json:"variants,omitempty"`
+	RoutingTemplates  map[string]routingTemplate `json:"routingTemplates,omitempty"`
+	normalizedWeights []uint32
+}
+
+type variant struct {
+	Resources []resource `json:"resources,omitempty"`
+	Weight    *uint32    `json:"weight,omitempty"`
+}
+
+type resource struct {
+	GVRShort  string  `json:"gvrShort"`
+	Name      string  `json:"name"`
+	Namespace *string `json:"namespace"`
+}
+
+type routingTemplate struct {
+	GVRShort string `json:"gvrShort"`
+	Template string `json:"template"`
+}
+
+// routemaps by their name
+type routemapsByName map[string]*routemap
+
+// routemaps contain every routemap known to Iter8
+type routemaps struct {
+	mutex sync.RWMutex
+	// map each namespace to its routemapsByName
+	nsRoutemap map[string]routemapsByName
+}
+
+/* types: end */
+
+const (
+	routemapStrSpec      = "strSpec"
+	weightAnnotation     = "iter8.tools/weight"
+	defaultVariantWeight = uint32(1)
+)
+
+// Weights provide the relative weights for traffic routing between variants
+func (s *routemap) Weights() []uint32 {
```

Comment:
What is the purpose of this function?  I am unclear on the use case to externalize it.

Reply:
This is how the `routemap templates` get weights for variants.

Reply:
Documented now.